java-racingcar-6 과제 제출합니다!

Hwangseoeun opened the pull request:
### 구현하지 못한 부분
1. 해당 기능에 대한 test 코드를 작성하지 못함
    - 선택한 자동차의 전진 카운팅 횟수를 증가시키는 기능 테스트
    - 차수별 실행 결과를 출력하는 기능 테스트
    - 최종 우승자의 이름을 출력하는 기능 테스트
    - 자동차 이름 5글자 이상인 경우에 대한 예외처리 테스트
    - 저장되어 있는 자동차가 없는 경우에 대한 예외처리 테스트
 
2. 최종 점검 과정에서 발견한 점
(코드 컨벤션 및 객체지향 생활 체조 원칙 기반으로 점검)
    - 'RacingService' 클래스 내 'printCountView' 메소드에 이중 for문 존재(객체지향 생활 체조 원칙을 지키지 못함)
    - 'RacingService' 클래스 내 'chooseRandomCars' 메소드에 같은 줄에 . 2개 이상 존재(객체지향 생활 체조 원칙을 지키지 못함)

### 구현하며 궁금했던 점
1. Random 값은 테스트 코드를 어떻게 작성해야 확인할 수 있는가?
    a. 전진하는 조건을 판단하는 기능 구현 과정에서 고민
        -> 무작위 값을 임의로 하나 정하는 기능
        -> 해당 값이 랜덤인지 아닌지 어떻게 확인해야할까..? 라는 고민이 생김
    b. 차수별 전진한 자동차를 기록하는 기능 구현 과정에서 고민
        -> n대의 자동차 중 일부 자동차를 랜덤으로 선택하는 기능
        -> 선택된 일부의 자동차가 랜덤인지 아닌지 어떻게 확인해야할까..? 라는 고민이 생김

Hwangseoeun committed:
docs: README 파일에 내용 추가
기능을 구현하기 전 기능 목록을 정리함
앞으로 사용할 커밋 컨벤션 종류를 정리함

Hwangseoeun committed:
docs: README 파일 수정
기존 프로젝트 루트에 위치한 README.md에 기능 목록 추가하였던걸 docs/README.md 위치로 수정함.

기존 프로젝트 루트의 README.md 파일에 추가했었던 기능 목록은 삭제하였음

README.md 파일에 들어가는 내용
1. 기능을 구현하기 전 기능 목록 정리

Hwangseoeun committed:
feat: Car 객체 생성 및 경주할 자동차의 이름을 입력받는 기능 구현

최대한 하나의 메서드에서 하나의 동작만 할 수 있도록 inputCarNames와 parseCarNamesToArray로 분리해서 구현함

1. inputCarNames
사용자가 경주할 자동차의 이름을 입력할 수 있도록 구현

2. parseCarNamesToArray
입력받은 자동차들의 이름을 파싱을 통해 배열에 저장하도록 구현

Hwangseoeun committed:
docs: 경주할 자동차의 이름을 입력받는 기능 구현 완료

Hwangseoeun committed:
test: 경주할 자동차의 이름을 입력받는 기능 테스트

1.inputCarNames
콘솔 입력을 시뮬레이션 하여 입력받는 값이 올바르게 들어오는지 테스트

2.parseCarNamesToArray
입력받은 문자열을 파싱하여 자동차의 이름 리스트를 올바르게 저장하는지 테스트
리스트의 순서와 상관없이 올바른 자동차 이름이 포함되어 있는지 검증

Hwangseoeun committed:
refactor: 클래스명 변경
'CarService'를 'RacingService'로 변경

Hwangseoeun committed:
refactor: 클래스명 변경
'CarServiceTest'를 'RacingServiceTest'로 변경

Hwangseoeun committed:
feat: 시도할 횟수를 입력받는 기능 구현

새로운 메서드 `InputNumberOfTries`를 추가하여 사용자가 시도할 횟수를 입력받도록 구현

Hwangseoeun committed:
docs: 시도할 횟수를 입력받는 기능 구현 완료

Hwangseoeun committed:
refactor: 메소드명 변경
카멜케이스를 적용하여 'InputNumberOfTries'을 'inputNumberOfTries'으로 변경

Hwangseoeun committed:
test: 시도할 횟수를 입력받는 기능 테스트
1.inputNumberOfTries
콘솔 입력을 시뮬레이션 하여 입력받는 값이 올바르게 들어오는지 테스트

Hwangseoeun committed:
feat: 무작위 값을 구하는 기능
0에서 9 사이에서 하나의 무작위 값을 구하도록 구현

Hwangseoeun committed:
feat: 무작위 값이 4 이상인지 판단하는 기능
0부터 9 사이에서 구한 하나의 무작위 값이 4 이상인지 boolean 값으로 return하도록 구현

Hwangseoeun committed:
refactor: 전진하는 조건을 판단하는 기능 수정

1.judgeRandomNumberSize
해당 메소드에 if문을 사용하지 않더라도 바로 return 값을 통해 판단할 수 있기에 수정

2.decideAdvance
'selectRandomNumber', 'judgeRandomNumberSize' 두 메소드를 연결하여 구현함

Hwangseoeun committed:
docs: 무작위 값을 구하는 기능 구현 완료

Hwangseoeun committed:
docs: 무작위 값이 4 이상인지 판단하는 기능 구현 완료

Hwangseoeun committed:
test: 무작위 값이 4 이상인지 판단하는 기능 테스트

1.judgeRandomNumberSize
임의의 값이 4 이상인지 판단 후 올바른 return을 하는지 테스트

Hwangseoeun committed:
refactor: carList 멤버변수 선언
Car 리스트를 재사용할 수 있도록 멤버 변수로 선언함

Hwangseoeun committed:
feat: n대의 자동차 중 일부 자동차를 랜덤으로 선택하는 기능 구현

1.selectRandomCarCount
몇대의 자동차를 뽑아낼 것인지 랜덤하게 결정

2.randomCarIndexNumbers
어떤 자동차를 선택할 것인지 랜덤한 인덱스 값을 통해 결정

3.chooseRandomCars
'randomCarIndexNumbers'에서 받은 랜덤한 인덱스 값을 통해 자동차의 이름들을 return하도록 구현

Hwangseoeun committed:
refactor: 경주할 자동차의 이름을 입력받는 기능 수정

1.saveCarNames
'inputCarNames', 'parseCarNamesToArray' 두 메소드를 연결하여 구현

Hwangseoeun committed:
feat: 선택한 자동차의 전진 카운팅 횟수를 증가시키는 기능 구현

1.findCarByName
'chooseRandomCars' 메서드의 return 값인 자동차 이름을 랜덤하게 저장한 배열을 이용하여, Car 객체를 조회

2.increaseCarCount
조회된 객체의 count 값을 증가시키도록 구현

Hwangseoeun committed:
refactor: Service 내 메서드 삭제 및 반환값 수정

1.saveCarNames
'inputCarNames', 'parseCarNamesToArray' 두 메서드를 연결하는 역할이였기에, Service 클래스에 있는 것보다는 Controller 클래스에 존재하는 것이 더욱 적합한 것 같아 이동함

2.getCarList
Car 객체의 정보를 조회하기 위해 생성

3.decideAdvance
기존에는 반환값을 설정하지 않았으나, 랜덤 값에 따른 전진 여부를 출력하기 위해 수정

Hwangseoeun committed:
refactor: Application 내 메서드 호출 삭제
Application 내에서 메서드를 호출하는 것보다 Controller 클래스를 생성하여 해당 파일에서 메서드를 호출하고 관리하는 것이 나을 것 같아 이동함

Hwangseoeun committed:
refactor: 메서드 호출 및 관리를 위해 추가
기존 'RacingService', 'Application' 내에 있던 메소드 호출 부분을 'RacingController' 클래스로 이동

Hwangseoeun committed:
docs: 차수별 전진한 자동차를 기록하는 기능 구현 완료

Hwangseoeun committed:
feat: 차수별 실행 결과를 출력하는 기능 구현
Car 객체에 저장되어 있는 값들을 조회한 후 count 값에 따라 '-' 추가되도록 구현

Hwangseoeun committed:
feat: 최종 우승자의 이름을 출력하는 기능 구현
우승자가 여러 명일 경우 ','를 이용하여 구분하여 출력되도록 구현함

Hwangseoeun committed:
docs: 최종 우승자의 이름을 출력하는 기능 구현 완료

Hwangseoeun committed:
refactor: 경주할 자동차의 이름을 입력받는 기능 수정
자동차 이름은 5자 이하만 가능하도록 조건 추가 및 예외처리

Hwangseoeun committed:
refactor: 상수 이름 대문자로 수정
기존 'maxCarNameLength'였던 상수 이름을 'MAX_CAR_NAME_LENGTH'로 수정

Hwangseoeun committed:
style: 띄어쓰기 한칸으로 통일되도록 수정

kelly6bf commented on docs/README.md:
구현할 프로그램의 기능 목록을 미리 생각해서 작성해보는 과정이 처음이라 쉽지 않았을텐데 잘 작성해주셨어요!
서은님이 작성할 프로그램의 큰 기능들이 한눈에 들어오네요 🤭

기능 목록을 문서로 미리 작성해보고 구현을 하는 경험은 어떠셨는지 궁금해요!

Hwangseoeun replied:
- 구현하며 느낀 점
사전에 작성해 둔 기능 목록을 참고하면서 구현하니 각 기능에 어떤 메서드들이 필요할지 파악할 때 도움이 됐어요!
그리고 기능 목록을 구체적으로 작성해두니 메서드를 어떻게 세부화하면 좋을지 파악하기 쉬웠던 것 같아요☺️

- 수정한 내용
'프리코스 2주차 웹 백엔드 피드백' 문서를 토대로 '쉼표를 기준으로 구분한다', '이름은 5자 이하만 가능하다' 등의 조건들이나 예외적인 상황들을 기능목록에 추가했어요!

kelly6bf commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,23 @@
+package racingcar;
+
+public class Car {
+    private final String carName;
```
```Car```라는 객체가 자신의 이름을 상태로 가지고 있다면 ```name```이라는 네이밍이 더 적절하지 않을까요?

- Car.carName
- Car.name

서은님은 어떤 네이밍이 더 와닿으시는것 같나요? 😋

Hwangseoeun replied:
Car.name 네이밍이 더 적절한 것 같아요!

구현을 할 당시에는 네이밍을 name으로 하게 될 경우 어떤 것의 name인지 헷갈릴 수도 있겠다고 생각이 들어서 carName을 선택했었어요🤔

그러나 커멘트를 보고 다시 생각해보니 말씀하신대로 name은 결국 Car 객체의 상태로 가지고 있는거니까 어떤 것의 name인지 헷갈릴 일이 없을 것 같네요!

수정하였습니댜~!

kelly6bf commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,23 @@
+
+public class Car {
+    private final String carName;
+    private int count;
```
자동차의 ```count```라는 상태는 어떤 개수를 의미할까요?
자동차에 타고 있는 사람의 수? 자동차가 가지고 있는 좌석의 수? 

단순 ```count```라는 이름은 어떤 개수를 의미하는지 빠르게 파악하기 힘든것 같아요. 자동차가 이동한 총 거리를 상태로 넣고 싶었다면 더 구체적인 이름을 고민해보는게 어떨까요?

(전 이런경우에 지선생님께 자문을 구해봅니다 🤭)

<img width="568" alt="image" src="https://github.com/user-attachments/assets/4d71b12f-ecbf-4ebf-804e-091086458ce0">

Hwangseoeun replied:
자동차의 count 상태는 자동차의 전진 횟수를 의미해요!

필드명을 통해 어떤 의미를 나타내는지 파악할 수 있도록 기존 'count'에서 'advanceCount'로 수정했습니다!

kelly6bf commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,23 @@
+
+    public Car(String carName) {
+        this.carName = carName;
+        this.count = 0;
```
자동차의 초기 위치는 0이기 때문에 생성자가 0으로 초기화 해주고 있네요!

하지만 모두가 '자동차의 초기 위치는 0부터 시작해!'라는 전제를 알고 코드를 확인할 수 있을까요? 누군가에게는 ```1```이 초기 값이라 생각할지도, 누군가는 ```-1```이라고 생각할 수 있기에 0이라는 숫자만 보고 자동차의 초기값이라는 지식을 빠르게 파악할 수 있을까요?

```매직 넘버```라는 개념을 찾아보고 개선해보면 좋을거 같습니다!
https://javabom.tistory.com/28

Hwangseoeun replied:
`매직 넘버`로 인해 값이 무엇을 의미하는지 빠르게 파악하기 어렵군요! 🤔

코드의 가독성을 높이고 유지 관리를 더욱 쉽게 할 수 있도록 `매직 넘버`가 아닌 `상수` 'DEFAULT_ADVANCE_COUNT'로 선언하도록 수정하였습니다!

kelly6bf commented on src/main/java/racingcar/RacingController.java:
```diff
@@ -0,0 +1,50 @@
+
+public class RacingController {
+
+    private RacingService racingService;
```
프로그램이 수행되면서 ```RacingService```의 인스턴스가 재할당될 여지가 있을까요?
만약 프로그램의 로직에 따라 초기에 설정된 ```RacingService```의 인스턴스가 재할당되는걸 의도하지 않았다면 원하지 않는 재할당을 막을 필요가 있지 않을까요? 이때는 ```final```을 활용해볼 수 있겠네요!

```java
private final RacingService racingService;
```

해당 필드 말고도 프로그램 전체적으로 재할당을 의도한 필드가 아닌것들 모두 마찬가지입니다!

Hwangseoeun replied:
RacingController 객체가 메모리에 존재하고 실제로 사용되는 동안 RacingService의 인스턴스는 재할당될 필요가 없는 것 같아요!

RacingService의 인스턴스가 `final`이 붙어있지 않다면, 예상치 못한 문제가 생길 수도 있겠네요.
예시) 동일한 객체를 참조할 것이라고 예상했으나, 다른 객체를 참조하고 있는 경우

재할당이 필요없는 필드들에 `final`을 붙여서 재할당을 막도록 수정하였습니다!

kelly6bf commented on src/main/java/racingcar/RacingController.java:
```diff
@@ -0,0 +1,50 @@
+    private RacingService racingService;
+
+    public RacingController() {
+        this.racingService = new RacingService();
```
생성자에서 직접 ```RacingService``` 인스턴스를 생성해서 할당해주고 있네요!

만약 테스트를 위해 임의로 만든 ```RacingService```를 만들고 싶다면... 지금 로직에서는 가능할까요?

Hwangseoeun replied:
지금 로직에서는 테스트를 위해 임의로 만든 RacingService가 아닌 RacingService의 실제 구현이 사용돼요!

테스트코드 작성을 고려해봤을 때, RacingController를 테스트 하는 경우 RacingService의 실제 동작을 검증하는 것은 불필요한 것 같아요 🤔

그렇기에 생성자에서 직접 RacingService 인스턴스를 생성해서 할당하는 것이 아닌, RacingService 객체를 생성자에서 주입받도록 해야겠네요!

수정하였습니댜~!

kelly6bf commented on src/main/java/racingcar/RacingService.java:
```diff
@@ -0,0 +1,132 @@
+public class RacingService {
+
+    private List<Car> carList;
+    private static final int MAX_CAR_NAME_LENGTH = 5;
```
제공된 java 컨벤션에 따르면 다음과 같이 필드 순서를 유지할 필요가 있겠네요!

```java
private static final int MAX_CAR_NAME_LENGTH = 5;

private List<Car> carList;
```

Hwangseoeun replied:
컨벤션을 토대로 상수, 멤버 변수, 생성자, 메서드 순으로 작성하도록 필드 순서를 수정하였습니다!

kelly6bf commented on src/main/java/racingcar/RacingService.java:
```diff
@@ -0,0 +1,132 @@
+    public String inputCarNames() {
+        System.out.println("경주할 자동차 이름을 입력하세요.(이름은 쉼표(,) 기준으로 구분)");
+
+        return Console.readLine();
```
사용자의 ```입력```을 받는 역할은 ```Service```의 책임일까요?

서은님이 생각하시는 ```Controller```와 ```Service```의 각 역할과 책임은 무엇인지 궁금합니다! 🤔

(Spring에서 보통 클라이언트의 API 호출을 통한 데이터 입력은 어떤 클래스가 책임을 수행했었죠? 👻)

Hwangseoeun replied:
Spring에서 데이터 입력은 Controller에서 수행했었네요..😖

코드를 작성할 때 제가 생각했던 Controller와 Service의 각 역할과 책임은

- Controller
    - 역할 : Service 단에서 생성된 메서드를 활용해 전반적인 흐름을 관리
    - 책임 : Service 단과 데이터를 주고 받으며, 사용자에게 결과를 보여주도록 전달해야 함
- Service
    - 역할 : 데이터 처리 및 객체의 동작을 관리
    - 책임 : 사용자로부터 입력받은 데이터를 처리하고, 적절한 결과를 생성. 해당 결과를 Controller에게 반환해야 함

이라고 생각했어요!

그러나 링크로 첨부해주신 mvc 패턴과 관련된 영상을 보고 다시 생각해보니 약간의 차이가 있는 것 같아요!

- Controller
    - 역할 : 사용자와 Service의 중간에서 데이터를 주고받을 수 있도록 하는 매개체, 전반적인 흐름을 관리
    - 책임 : 
        1. 사용자로부터 입력을 받아 적절한 형식으로 변환하고, 이를 Service 계층에 전달
        2. Service 계층에서 받은 결과를 가공하여, 사용자에게 필요한 형태로 반환
        3. 애플리케이션의 흐름을 제어하고, 사용자 요청에 따라 적절한 비즈니스 로직이 실행되도록 함
- Service
    - 역할 : 데이터 처리 및 객체의 동작을 관리
    - 책임 : Controller로부터 받은 사용자의 데이터를 처리하고, 적절한 결과를 생성. 해당 결과를 다시 Controller에게 반환해야 함

이렇게 이해했어요!

이를 적용하여 수정하자면, Controller에서 사용자의 입력을 받아 Service 단으로 데이터를 넘겨야 할 것 같아요. 이후 데이터 처리를 하여 다시 Controller로 넘겨 사용자에게 데이터를 보여줄 수 있도록 구현해야 할 것 같네요..!!

혹시 성욱님은 Controller와 Service의 각 역할과 책임이 무엇이라 생각하시나요?? 😋

kelly6bf commented on src/main/java/racingcar/RacingService.java:
```diff
@@ -0,0 +1,132 @@
+                sb.append("-");
+            }
+
+            System.out.println(carName + " : " + sb.toString());
```
사용자에게 무언가를 ```출력```하는 행위 역시 어떤 객체가 가지면 좋을 책임일까요? 위 입력에 대한 코멘트와 함께 생각해보면 좋을거 같아요! 🤔

Hwangseoeun replied:
입력하는 행위와 출력하는 행위 모두 사용자와 직접적으로 데이터나 값을 주고받아야 하기에, Service보다는 Controller의 책임이 더 적절한 것 같아요!

kelly6bf commented on src/main/java/racingcar/RacingService.java:
```diff
@@ -0,0 +1,132 @@
+        }
+
+        return carList.stream()
+                .filter(car -> car.getCount() == maxCount)
```
자동차의 현재 이동 거리가 ```macCount```와 동일한지 판단하는 책임 역시 누구에게 가는게 좋을까요?


??? : 서은님은 성이 뭐에요?
서은 : 저는 황씨입니다!
??? : 서은님은 황씨가 맞군요!
서은 : 그냥 나한테 물어보면되는거 아닌가...?

어떤 행위에 대한 책임은 그 행위에 필요한 지식을 가지고 있는 객체가 가장 잘하겠죠!

특정 자동차의 이동 거리가 macCount인지 판단하는 행위에 필요한 지식이 ```특정 자동차의 이동 거리```라고 한다면 그 지식을 가지고 있는 객체가 수행하는게 적절할 수 있겠죠! 그건 누구일까요? 🤔

Hwangseoeun replied:
Car 객체에 advanceCount 상태가 존재하기에, `특정 자동차의 이동 거리` 지식을 가지고 있는 객체는 Car 객체에요!
Car 객체에게 책임이 가는게 좋겠네요 🤔

그렇다면

1. 'increaseCarCount' : 전진한 자동차의 advanceCount 값을 증가시키는 메서드
2. 'findCarByName' : 자동차의 이름을 통해 자동차의 정보를 불러오는 메서드 
3. 'findMaxCount' : 모든 자동차들의 이동거리 중 최대값을 조회하는 메서드

Car 객체가 해당 행위들에 대한 지식을 가지고 있는데, 이에 대한 책임도 Car 객체가 가지는게 맞을까요?? 🥹

kelly6bf commented on src/main/java/racingcar/RacingController.java:
```diff
@@ -0,0 +1,50 @@
+        this.racingService = new RacingService();
+    }
+
+    public void start() {
```
모든 메서드가 ```public```으로 오픈되어 있네요 🤔 해당 객체를 사용하는 클라이언트에서 이 객체의 모든 메서드에 자유롭게 접근할 필요가 있을까요?

다른 객체의 메서드 역시 생각해보면 좋을거 같아요!

Hwangseoeun replied:
모든 메서드에 자유롭게 접근할 필요 없이 각 객체의 역할과 책임에 따라 접근 권한이 달라질 것 같아요!

링크로 첨부해주신 public, private 관련된 게시글을 통해 
'이 클래스를 사용하는 객체에게 A ~ Z까지 메소드 중 어느 Method를 쓰는게 다른 Method들에 비해 안정적인지 미리 알려주기 위해 public / private 를 결정한다. 그래서 테스트를 작성할때도 세부구현을 담당하는 private interface (private Method)를 테스트할게 아니라, 쉽게 변경되지 않고 다른 객체가 안정적으로 의존할 수 있는 public interface (public Method) 를 대상으로 삼아야한다.' 
라는 점을 새롭게 알 수 있었어요!

해당 메소드가 어디까지 접근 권한이 있는지 파악한 후 그에 맞춰 수정하였습니다!!

kelly6bf reviewed (changes requested):
안녕하세요 서은님! 미션 진행하시느라 정말 고생 많으셨습니다!

기능 목록 정리부터 객체간 책임 나누기, 테스트 코드 작성에 컨벤션 준수까지... 처음 시도해보는 것들이 많아 많이 어색하고 막막했을게 눈에 보이네요... 🥹

단순히 돌아가는 기능을 구현하는걸 떠나 이런 다양한 제약 조건들을 생각하면서 진행해본 경험은 어떠셨나요? 굳이 이렇게 까지 해야하나 싶을 정도의 제약 사항을 지킴으로써 무언가 이점을 얻었다는 느낌이 드셨나요?

항상 무엇을 시도하던 그것이 주는 이점이 무엇인지를 생각해보며 진행한다면 더 재밌게 즐길 수 있지 않을까 생각이 드네요!

간단하게 리뷰를 남겼으니 확인하시고 코드 개선 해주시면 됩니다! 리펙토링 역시 처음이라 많이 어려우시겠지만 차근차근 진행해봐요 😁

궁금한건 언제든 물어보세요!

## 고민해봤으면 하는 부분들
- 객체 A와 B가 있을때, 어떤 행위에 대한 책임을 누군가에게 줘야 한다면... 어떤 기준으로 할당할 수 있을까?
- Controller? Service? Domain? 뭔가 들어봤던 이 객체들은.. 각각이 어떤 책임을 가질까?

kelly6bf reviewed (changes requested):
안녕하세요 서은님! 미션 진행하시느라 정말 고생 많으셨습니다!

기능 목록 정리부터 객체간 책임 나누기, 테스트 코드 작성에 컨벤션 준수까지... 처음 시도해보는 것들이 많아 많이 어색하고 막막했을게 눈에 보이네요... 🥹

단순히 돌아가는 기능을 구현하는걸 떠나 이런 다양한 제약 조건들을 생각하면서 진행해본 경험은 어떠셨나요? 굳이 이렇게 까지 해야하나 싶을 정도의 제약 사항을 지킴으로써 무언가 이점을 얻었다는 느낌이 드셨나요?

항상 무엇을 시도하던 그것이 주는 이점이 무엇인지를 생각해보며 진행한다면 더 재밌게 즐길 수 있지 않을까 생각이 드네요!

간단하게 리뷰를 남겼으니 확인하시고 코드 개선 해주시면 됩니다! 리펙토링 역시 처음이라 많이 어려우시겠지만 차근차근 진행해봐요 😁

궁금한건 언제든 물어보세요!

## 고민해봤으면 하는 부분들
- 객체 A와 B가 있을때, 어떤 행위에 대한 책임을 누군가에게 줘야 한다면... 어떤 기준으로 할당할 수 있을까?
- Controller? Service? Domain? 뭔가 들어봤던 이 객체들은.. 각각이 어떤 책임을 가질까?
https://www.youtube.com/watch?v=ogaXW6KPc8I&t=456s
- public? private? 언제 무엇을 사용해야할까?
https://jojoldu.tistory.com/588

Hwangseoeun committed:
refactor: Car 객체 필드명 수정
1.carName
기존 'carName'이었던 필드명을 'name'으로 수정

2.count
기존 'count'이었던 필드명을 'advanceCount'로 수정

Hwangseoeun committed:
refactor: Car 객체의 advanceCount 초기값 상수 선언
기존 advanceCount의 초기값은 매직 넘버인 0으로 선언되어있었음
해당 값이 무엇을 의미하는지 빠르게 파악할 수 있도록 상수 'DEFAULT_ADVANCE_COUNT'로 초기값 선언

Hwangseoeun committed:
refactor: 필드 순서 변경
생성자보다 상수가 우선적으로 작성되도록 필드 순서 수정

Hwangseoeun committed:
docs: 기능 목록 수정
기능 요구 사항에 적혀있는 세부적인 조건들 추가

Hwangseoeun committed:
refactor: 재할당 막기 위해 final 추가

Hwangseoeun committed:
refactor: RacingService 객체 의존성 주입
RacingController 생성자에서 직접 RacingService 인스턴스를 생성해서 할당하는 것이 아닌, RacingService 객체를 생성자에서 주입받도록 수정

Hwangseoeun committed:
refactor: 재할당 막기 위해 final 추가

Hwangseoeun committed:
refactor: 입력, 출력 메서드 Controller로 이동
입력하는 행위와 출력하는 행위 모두 사용자와 직접적으로 데이터나 값을 주고받아야 하기에, Service에서 Controller로 이동

Hwangseoeun committed:
refactor: 사용하지 않는 외부 의존성 제거

Hwangseoeun committed:
test: Service 내부에 있던 메서드를 Controller로 옮김으로써 테스트 코드도 동일하게 수정

Hwangseoeun committed:
refactor: 각 메서드 접근 제한자 수정 및 메서드 순서 변경

모든 메서드의 접근 제한자는 public이었으나, 어디까지 접근할지에 따라 public, protected, private, default로 선언하도록 수정
